modules/notify/src/main/NotifyApi.scala

```diff
@@ -28,16 +29,53 @@ final class NotifyApi(bus: lila.common.Bus, repo: NotificationRepo) {
   def addNotification(notification: Notification): Funit = {
 
     // Add to database and then notify any connected clients of the new notification
-    repo.insert(notification) >>-
-      getUnseenNotificationCount(notification.notifies).
-        map(NewNotification(notification, _)).
-        foreach(notifyConnectedClients)
+    insertOrDiscardNotification(notification) map {
+      _ match {
+        case None => funit
+        case Some(notif) =>
+            getUnseenNotificationCount(notif.notifies).
+              map(NewNotification(notif, _)).
+              foreach(notifyConnectedClients)
+      }
+    }
   }
 
   def addNotifications(notifications: List[Notification]) : Funit = {
     notifications.map(addNotification).sequenceFu.void
   }
 
+  /**
+    * Inserts notification into the repository.
+    *
+    * If the user already has an unread notification on the topic, discard it.
+    *
+    * If the user does not already have an unread notification on the topic, returns it unmodified.
+    */
+  private def insertOrDiscardNotification(notification: Notification): Fu[Option[Notification]] = {
+    val recentNotifications = getNotifications(notification.notifies, 1, 15)
```

Comment:
how about filtering directly in mongodb and only fetching the relevant ones, if any.